### ISSUE TITLE
Try to get a units converter for a type before falling back to built-in behavior

### ISSUE THREAD

**Issue by ryan-gunderson** [matplotlib/matplotlib#20502]
When given an object to plot, should matplotlib attempt to use a registered `units.ConversionInterface` for that object's type (and mro) before falling back to other built-in defaults?

This question/suggestion prompted by the following example:

Say one has a class
```
class DateTimeNano(np.ndarray):
    ...
```
subclassing `np.ndarray` with the convention that `DateTimeNano` stores nano-seconds since epoch.  When plotting a `DateTimeNano`, one may want to make use of the built in utilities for date tick location/formatting.  If one writes a `units.ConversionInterface` like
```
class DateTimeNanoConverter(units.ConversionInterface):
    NANOS_PER_DAY = 86_400*1_000_000_000
    @staticmethod
    def convert(value, unit, axis):
        # Values should be in floating point days since epoch to play nice with mpl
        day_floats = value / DateTimeNanoConverter.NANOS_PER_DAY 
        return day_floats

    @staticmethod
    def axisinfo(unit, axis):
        majloc = dates.AutoDateLocator()
        majfmt = dates.AutoDateFormatter(majloc)
        return AxisInfo(majloc=majloc, majfmt=majfmt, label='date time')

    @staticmethod
    def default_units(x, axis):
            return 'date_time'
```
registers it with
```
units.registry[DateTimeNano] = DateTimeNanoConverter
```
and then calls
```
plt.plot(my_date_time_nano, my_y_values)
```
only `DateTimeNanoConverter.default_units` and `DateTimeNanoConverter.axisinfo` are used.  `DateTimeNanoConverter.convert` is skipped entirely as [`units._is_natively_supported`](https://github.com/matplotlib/matplotlib/blob/b7ce415c15eb39b026a097a2865da73fbcf15c9c/lib/matplotlib/units.py#L56) recognizes that it is an iterable and so a converter for the type `DateTimeNano` is never searched for (it also appears that this would be short circuited in [`units.Registry.get_converter`](https://github.com/matplotlib/matplotlib/blob/b7ce415c15eb39b026a097a2865da73fbcf15c9c/lib/matplotlib/units.py#L181)).

If the step of searching for a converter by iterating over the type's mro were moved up in the order of trying to get values out of an object, this problem would not occur.  It also seems like an overall preferred behavior as searching for a way to get an array of values out (using a converter or built-in code paths) should always start with the most specific implementation and work out.
 This does not currently happen if types happen to have a particular inheritance structure or set of attributes.

**Comment by jklymak** [comment]
My understanding is that the units registry works on the data type of the _first_ element of the iterable, not the iterable.  i.e. `my_date_time_nano[0]`.    So if all your iterables are floats, matplotlib just assumes they are native.  I don't think there is a way to get the units machinery to check the class of the container with the current registry, but it is something to consider for the future.

**Comment by ryan-gunderson** [comment]
@jklymak That is the current behavior for iterables, yes.  For objects not caught by the handful of built-in cases, the converter is used on the object.  For example:
```
class Foo:
    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = c
        return

class Foo:
    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = c
        return

class FooConverter(units.ConversionInterface):
    @staticmethod
    def convert(value, unit, axis):
        xs = [value.a, value.b, value.c]
        return xs

units.registry[Foo] = FooConverter
```
During a call to `plt.scatter(my_foo, my_other_foo)`, `FooConverter.convert` is called resulting in plotting three points (however this does not work for `plt.plot` as the `value` argument to the `convert` function is an `ndarray` in that case; I'm not familiar with why).

Perhaps this is an abuse of the units system and so it may not make sense as part of the units module, but it seems reasonable to be able to, given an arbitrary object, register how it would like to be interpreted for plotting.

**Comment by jklymak** [comment]
I think you may want to create a custom dtype. https://stackoverflow.com/questions/2350072/custom-data-types-in-numpy-arrays

**Comment by jklymak** [comment]
More to your immediate concern, matplotlib can handle nanoseconds from an epoch already using datetime64.  Right now our default epoch is 1970-01-01, and nanoseconds in datetime64 have +/-292y of usable range.  So unless you plan to do nanoseconds for more than 584 years, something like `datetime = nano_seconds.astype('datetime64[ns]')`should work with the default locators and formatters.

**Comment by ryan-gunderson** [comment]
I agree that casting to `datetime64` can solve that particular example.  It seems that the units module is not intended to solve the problem of making arbitrary things plot-able.  So a separate issue would be appropriate for discussing that.  Thanks!

**Comment by tacaswell** [comment]
It looks the behavior of looking at `_is_natively_supported` was added in 3.2 via https://github.com/matplotlib/matplotlib/pull/13738 which makes it so that we do not get to https://github.com/matplotlib/matplotlib/blob/4f2f4debcd05ac72b6c3051b5d1efc2643a87fc2/lib/matplotlib/units.py#L193-L197 .  I think the changes in #13738 inadvertently broke the functionality that @ryan-gunderson is looking for.

**Comment by ryan-gunderson** [comment]
Perhaps moving the transversal of the mro to the front would make sense?  But I am still unsure of whether or not the units module is intended to handle coercing arbitrary objects into arrays-likes for plotting or just coercing unrecognized scalars to numbers.
